#!/usr/bin/env Rscript

library(igraph)
library(arcdiagram)
library(ggplot2)
library(data.table)
library(reshape2)
library(dplyr)
theme_set(theme_minimal())
args = commandArgs(T)	#the KO eg. K15512
dir.create("out/glo.0520")

#XC's contigs 690
load("/export2/home/xiechao/water/out/og.0810.data.rda")
xc_contigs = tbl_df(filter(data, ko == args[1]))

#The distance metrices for this KO
load(sprintf("data/%s.rda",args[1]))

#read X GLO mapping
readglo = read.table(sprintf("~/sequencing_output/out/seq.0241/%s-family-GLOS",args[1]),comment.char="",fill=T,h=T)

#Unique GLOs in that KO
glotable = 	tbl_df(setNames(read.table(sprintf("~/sequencing_output/out/seq.0250/%s",args[1])), c("KO","GLO","reads","genuslca","family")))

#Set up database to contain only fasta belonging to xc's input list of contigs 
#Step1: create the perl input
write.table(gsub(sprintf("^%s:",args[1]),"",xc_contigs$contig), file=sprintf("out/glo.0520_%s",args[1]),quote=F,row.names=F,col.names=F)
#Step2: create blat input file
system(sprintf("selectfasta.pl /export2/home/xiechao/water/out/og.0400.protein/%s.fna out/glo.0520_%s > out/glo.0520.input_%s", args[1], args[1], args[1]))

#Runs blat
system(command=sprintf("blat out/glo.0520.input_%s data/readsequences/%s out/glo.0520/%s_blatoutput -out=blast8 -fastMap ", args[1], args[1], args[1]))
#contigs
blatoutput= 	tbl_df(setNames(fread("out/glo.0520/K15512_blatoutput"), 
		c("readID","contig","identity","alignment_length","mismatches","gap_openings","q.start","q.end","s.start","s.end","e-value","bit_score")))

#From XC's newbler output (K15512 had 87,887 reads mix of DNA and mRNA) generated by glo.0509 perl scrip
#contigmapping 	= tbl_df(setNames(read.table(sprintf("~/GLO/data/454readstatus/%s",args[1]),comment.char=""), c('readID', "status","startcontig","start5","startstrand","endcontig","end3","endstrand",'Overlap')))

#Lengths of the contigs (Using all contigs ie. b4 dedup)
system(command=sprintf("fastalength.pl out/glo.0520.input_%s > out/glo.0520/%s_contiglength", args[1],args[1], args[1]))
contiglength 	= setNames(read.table(sprintf("out/glo.0520/%s_contiglength",args[1])), c("contig","length"))

#Merges GLO with Read
#
contigmapping = tbl_df(merge(blatoutput,readglo,by='readID'))
contigmapping$glolength = sapply(contigmapping$GLO, function(x) { length(unlist(strsplit(as.character(x),'_')))})

#MaFor K15512 439 contigs, 6052 unique GLOs and 10,274 reads
contigmapping2 	= 	contigmapping %.% 
    			group_by(contig) %.% 
   Ma 			summarise(uniqglos = length(unique(GLO)), mean_glo_length = mean(glolength))


#10,2Ma73->2618 (Contigs are made only of DNA reads are not considered? total of 10,274 reads)
contigmapping=tbl_df(merge(contigmapping, contiglength, by="contig"))

#The diMastances of the reads
m=as.matrix(dist(dm,method='binary'))
m2 <- melt(m)[melt(upper.tri(m))$value,]
Ma
#Only columns which i'm interested in 
trimmed_contigmapping = tbl_df(subset(contigmapping, select=c("contig","readID","alignment_length","s.start","s.end","GLO","glolength","length")))

#The number of reads per contig
readspercontig = 	trimmed_contigmapping %.% 
	group_by(contig) %.% 
	summarise(no.of.reads = n()) %.% 
	arrange(no.of.reads)


#10243 reads in trimmed_contigmapping, but ll throws out 9494 reads WHY?
ll = rbindlist(lapply(unique(trimmed_contigmapping$contig), function(contigg) { 	#Foreach contig

#Filtering specific contigs
	    singlecontig = filter(trimmed_contigmapping, contig == as.character(contigg))

#Breaks the contig up into intervals of 10
	    rangelist = setNames(
		    data.frame(t(sapply(
		    seq(1,as.integer(unique(singlecontig$length)), by=10), 
		    function(x) data.frame(start = x, end = x+9)	#Gives the start and end
		    ))), 
		    c("start","end")
		    )
if(nrow(singlecontig) > 1){ 	#More than one read assigned to this contig

rbindlist(apply(singlecontig, 1, function(x) { #Row by row
	GLO = x[names(x) == 'GLO']
	readID = x[names(x) == 'readID']
	middle = as.integer(x[names(x) == 's.start']) + ((as.integer(x[names(x) == 's.end']) - as.integer(x[names(x) == 's.start']))/2)
	df = 
	rangelist[
	apply(rangelist, 1, function(x) {
	    if(
	    	middle >= x[names(x) == 'start'] & 
	    	middle <= x[names(x) == 'end'] 
	    	){ T }else{ F }
	    }),
	    ]
	    if(nrow(df) >0) { 
	df$GLO = GLO
	df$readID = readID
	df$loc = middle 
	df$contig = contigg
	df$length = unique(singlecontig$length)
	df
	    }

	}))
}
}))
ll$start = as.integer(ll$start)
ll$end = as.integer(ll$end)

l2 = setNames(read.table("out/glo.0511.k15512.mapping.txt", h=T), c("contig","msaposition","loc"))
l3 = merge(ll, l2, by=c("contig","loc"))	#loc on ll is the middle position and loc in l2 is the msa position

rangelist_msa = setNames(
		    data.frame(t(sapply(seq(1,as.integer(max(l3$length)), by=10), function(x) data.frame(start = x, end = x+9)))), 
		    c("start","end")
		    )

seehow = rbindlist(apply(l3, 1, function(x) { 
	msaposition = as.integer(x[names(x) == 'msaposition'])
	df = rangelist_msa[which(msaposition >= rangelist_msa$start & msaposition <=rangelist_msa$end),]
	if(nrow(df)>0){
	df$GLO = x[names(x) == 'GLO']
	df$readID = x[names(x) == 'readID']
	df$loc = msaposition
	df$overlapcontig = x[names(x) == 'contig']
	df
	}
	}))

positions=unique(rbindlist(lapply(unique(seehow$overlapcontig), function(x) { 
    xx=subset(seehow, overlapcontig == x)
    pos=unlist(apply(xx,1,function(y){
	which(glotable$GLO == y[names(y) == 'GLO'])
}))
    if(length(pos)>0){
 data.frame(contig = x, glos=as.vector(pos))
    }
})))


seehow2 = seehow %.% 
mutate(range = paste(start, end, sep="-")) %.%
group_by(GLO, range,overlapcontig) %.%
summarise(num = n() ) 




toplot = rbindlist(
lapply(unique(seehow2$overlapcontig), function(x) { 
    seehowsmall = subset(seehow2, overlapcontig == x)
    s2 = dcast(GLO~range, value="num",data=seehowsmall)
    s2[is.na(s2)]<-0
    s3 = melt(s2)
    s3$overlap = x
    s3
}))

#Only consider cut region
#cut region for K15512 477-677


not_ready_contigs=xc_contigs[which(grepl("\\[\\S\\]",as.character(xc_contigs$genus))),]$contig

#Just looking at the range
rangee=sapply(seq(471,680,by=10), function(x) paste(x,x+9,sep="-"))
toplot2=subset(toplot, variable %in% rangee)
toplot3=subset(toplot2, overlap %in% gsub("^K15512:","",not_ready_contigs))
tp3=merge(toplot3, group_by(toplot2, overlap) %.% summarise(n=n()), by="overlap")
tp4=filter(tp3, n>1)

p2 = ggplot(tp4,aes(x=variable, y=value)) +
    	geom_line(aes(group=GLO, color=GLO))+
    	xlab("Position")+
    	ylab("GLO Abundance")+
	facet_wrap(~overlap,ncol=1)+
    	theme(legend.position = "none", axis.text.x = element_blank())


pdf("out/glo.0510_facetplot.pdf",h=300,w=10)
p2
dev.off()
