#!/usr/bin/env Rscript

library(igraph)
library(arcdiagram)
library(dplyr)
library(ggplot2)
library(data.table)
library(reshape2)
theme_set(theme_minimal())
args = commandArgs(T)

#Needs input from 0511 for the coordinates

#The distance metrices for this KO
load(sprintf("data/%s.rda",args[1]))

#The read <-> GLO mapping
readglo 	= tbl_df(read.table(sprintf("~/sequencing_output/out/seq.0241/%s-family-GLOS",args[1]),comment.char="",fill=T,h=T))

#From XC's newbler output (K15512 had 87,887 reads mix of DNA and mRNA) generated by glo.0509 perl scrip
contigmapping 	= tbl_df(setNames(read.table(sprintf("~/GLO/data/454readstatus/%s",args[1]),comment.char=""), c('readID', "status","startcontig","start5","startstrand","endcontig","end3","endstrand",'Overlap')))

#Lengths of the contigs (Problematic)
contiglength 	= tbl_df(setNames(read.table(sprintf("out/glo.0509/contiglength/%s",args[1])), c("Overlap","contig","length")))

#Unique GLOs in that KO
glotable = 	tbl_df(setNames(read.table(sprintf("~/sequencing_output/out/seq.0250/%s",args[1])), c("KO","GLO","reads","genuslca","family")))

#Merges GLO with Read
#In the case of K15512, only 10,274 out of 15,086 of the reads were mappable to the newbler contigs
contigmapping = tbl_df(merge(contigmapping,readglo,by='readID'))
contigmapping$glolength = sapply(contigmapping$GLO, function(x) { length(unlist(strsplit(as.character(x),'_')))})

#For K15512 439 contigs, 6052 unique GLOs and 10,274 reads
contigmapping2 	= 	contigmapping %.% 
    			group_by(startcontig) %.% 
    			summarise(uniqglos = length(unique(GLO)), mean_glo_length = mean(glolength))

#Plotting the number of unique GLOs against the length of the GLOs###################################
pdf(sprintf("out/glo.0510/%s_uniqueGLOs.per.contig.pdf", args[1]))
qplot(uniqglos, mean_glo_length , data = contigmapping2)+
	scale_x_log10()+
    	xlab("# unique GLOs")+
    	ylab("Average length of GLOs")
dev.off()
####################################################################################################

colnames(contigmapping)[colnames(contigmapping) == 'startcontig']<-'contig'

#10,273->2618 (Contigs are made only of DNA reads are not considered? total of 10,274 reads)
#Unique combination: 1571 in contigmapping, 682 in contiglength hmmm 
#Only used dedup contigs
contigmapping=tbl_df(merge(contigmapping, contiglength, by=c("Overlap","contig")))

#The distances of the reads
m=as.matrix(dist(dm,method='binary'))
m2 <- melt(m)[melt(upper.tri(m))$value,]

trimmed_contigmapping = cbind(
select(contigmapping, c(Overlap:readID, GLO:length)), 
do.call(rbind,apply(contigmapping, 1, function(x) { 
startend=sort(c(as.integer(x[names(x) == 'start5']),as.integer(x[names(x) == 'end3'])))
    data.frame(
    start=startend[[1]], 
    end=startend[[2]]
    )
})))

trimmed_contigmapping = trimmed_contigmapping %.% 
	mutate(overlapcontig = paste(Overlap, contig,sep=""))

#The number of reads per contig
readspercontig = 	trimmed_contigmapping %.% 
	group_by(overlapcontig) %.% 
	summarise(no.of.reads = n()) %.% 
	arrange(no.of.reads)


#Arc plots
lapply(unique(trimmed_contigmapping$Overlap), function(overlap) { 
    pdf(sprintf("out/glo.0510/%s.pdf",overlap),w=10,h=5)
    lapply(unique(subset(trimmed_contigmapping, Overlap == overlap)$contig), function(contigg){ 
    singlecontig = filter(trimmed_contigmapping, Overlap == as.character(overlap), contig == as.character(contigg))


if(unique(singlecontig$length) > 200) {
rangelist = setNames(
		    data.frame(t(sapply(seq(1,as.integer(unique(singlecontig$length)), by=10), function(x) data.frame(start = x, end = x+99)))), 
		    c("start","end")
		    )


gloloc = do.call(rbind,(apply(singlecontig, 1, function(x) { 
	GLO = x[names(x) == 'GLO']
	readID = x[names(x) == 'readID']
	middle = as.integer(x[names(x) == 'start']) + ((as.integer(x[names(x) == 'end']) - as.integer(x[names(x) == 'start']))/2)
	df = 
	rangelist[apply(rangelist, 1, function(x) {
	    if(middle >= x[names(x) == 'start'] & middle <= x[names(x) == 'end'] ){ T }else{ F }
	    }),]
	    if(nrow(df) >0) { 
	df$GLO = GLO
	df$readID = readID
	df$loc = middle 
	df
	    }
	})))

glotable$num = 1:nrow(glotable)

vertex.details	= merge(gloloc, select(glotable, c(GLO,num)), by='GLO')
vertex.details$node = paste(vertex.details$start, vertex.details$end, sep="-")

#Get the relavant distances

    vertices = vertex.details %.% 
    group_by(node) %.% 
    summarise(readnum = n())

vertices = vertices[match(apply(rangelist, 1, function(x) paste(x[1],x[2],sep="-")), vertices$node),]
vertices = vertices[complete.cases(vertices),]


if(length(unique(vertex.details$node)) > 2) { 

#For all possible pairs of the nodes, calculate the distance 
edgelist = as.matrix(subset(
do.call(rbind,
apply(t(combn(unique(vertex.details$node), 2)), 1, function(x) {

#GLOs in that node
	pair1 = subset(vertex.details, node == x[1])$num
	pair2 = subset(vertex.details, node == x[2])$num

	data.frame(
	node1=x[1],
	node2=x[2],
	shared = nrow(subset(unique(melt(m[pair1,pair2])), value < 0.2))
	)
})), shared > 0))

if(nrow(edgelist) > 0) { 
g = graph.data.frame(edgelist, vertices = vertices,directed =F)

arcplot(get.edgelist(g),vertices = vertices$node, lwd.arcs = E(g)$shared, cex.nodes = sqrt(V(g)$readnum), main = sprintf("Overlap %s Contig %s",overlap, contigg))


gloloc2 = gloloc %.% 
mutate(range = paste(start, end, sep="-")) %.% 
group_by(GLO, range) %.%
summarise(num = n())

gloloc3 = dcast(GLO~range, data=gloloc2)
gloloc3[is.na(gloloc3)]<-0
gloloc4 = melt(gloloc3)
p = ggplot(gloloc4, aes(x=variable, y=value))+ 
    geom_line(aes(group=GLO, color=GLO))+
    ggtitle(paste(overlap, contigg, sep=" "))+
    xlab("Position")+
    ylab("GLO Abundance")+
    theme(legend.position = "none", axis.text.x = element_blank())

pdf(sprintf("out/glo.0510/tfplots/%s_%s_tfplot.pdf",overlap, contigg ),w=10,h=10);
print(p)
dev.off()
    }}}
    })
   dev.off() 
    })

ll = rbindlist(lapply(unique(trimmed_contigmapping$Overlap), function(overlap) { 
    rbindlist(lapply(unique(subset(trimmed_contigmapping, Overlap == overlap)$contig), function(contigg){ 
    singlecontig = filter(trimmed_contigmapping, Overlap == as.character(overlap), contig == as.character(contigg))

gloloc = do.call(rbind,(apply(singlecontig, 1, function(x) { 
	GLO = x[names(x) == 'GLO']
	readID = x[names(x) == 'readID']
	middle = as.integer(x[names(x) == 'start']) + ((as.integer(x[names(x) == 'end']) - as.integer(x[names(x) == 'start']))/2)
	df = 
	rangelist[apply(rangelist, 1, function(x) {
	    if(middle >= x[names(x) == 'start'] & middle <= x[names(x) == 'end'] ){ T }else{ F }
	    }),]
	    if(nrow(df) >0) { 
	df$GLO = GLO
	df$readID = readID
	df$loc = middle 
	df$contig = contigg
	df$overlap = overlap
	df$length = unique(singlecontig$length)
	df
	    }
	})))
gloloc
    }))	})	)

ll$start = as.integer(ll$start)
ll$end = as.integer(ll$end)
#write.table(x=ll, file="out/glo.0510.readlocs.txt",quote=F,row.names=F, sep="\t")

l2 = setNames(read.table("out/glo.0511.k15512.mapping.txt", h=T), c("overlap","contig","msaposition","loc"))
l3 = merge(ll, l2, by=c("overlap","contig","loc"))	#loc on ll is the middle position and loc in l2 is the msa position
l3$overlapcontig = with(l3, paste(overlap, contig, sep=""))

rangelist_msa = setNames(
		    data.frame(t(sapply(seq(1,as.integer(max(l3$msaposition)), by=10), function(x) data.frame(start = x, end = x+99)))), 
		    c("start","end")
		    )

seehow = 
rbindlist(apply(l3, 1, function(x) { 
	msaposition = as.integer(x[names(x) == 'msaposition'])
	df = rangelist_msa[which(msaposition >= rangelist_msa$start & msaposition <=rangelist_msa$end),]
	df$GLO = x[names(x) == 'GLO']
	df$readID = x[names(x) == 'readID']
	df$loc = msaposition
	df$overlapcontig = x[names(x) == 'overlapcontig']
	df
	}))

seehow2 = seehow %.% 
mutate(range = paste(start, end, sep="-")) %.%
group_by(GLO, range,overlapcontig) %.%
summarise(num = n() ) 

toplot = rbindlist(
lapply(unique(seehow2$overlapcontig), function(x) { 
    seehowsmall = subset(seehow2, overlapcontig == x)
    s2 = dcast(GLO~range, value="num",data=seehowsmall)
    s2[is.na(s2)]<-0
    s3 = melt(s2)
    s3$overlap = x
    s3
}))

p2 = ggplot(toplot,aes(x=variable, y=value)) + 
    	geom_line(aes(group=GLO, color=GLO))+
    	xlab("Position")+
    	ylab("GLO Abundance")+
	facet_wrap(~overlap,ncol=1)+
    	theme(legend.position = "none", axis.text.x = element_blank())
ggsave(
pdf("iwannadie.pdf",h=300,w=10)
p2
dev.off()
